### PR TITLE
feat(executor): added new encoders for numeric data

### PR DESCRIPTION
**Pull request by anish2197**
- Added TSNE : An alternative for PCA. Uses KL divergence as a loss.
https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html

- Added Feature Agglomeration, Random Gaussian Projection and Random Sparse Projections : Computationally inexpensive alternatives for PCA and TSNE.
https://scikit-learn.org/stable/modules/generated/sklearn.cluster.FeatureAgglomeration.html
https://scikit-learn.org/stable/modules/generated/sklearn.random_projection.GaussianRandomProjection.html
https://scikit-learn.org/stable/modules/generated/sklearn.random_projection.SparseRandomProjection.html

### CODE REVIEW

**Comment by anish2197** [comment]
> I'm aware that the best practice in PEP is to put every class variables in `__init__`. However, in Jina, everything in `__init__` will later be serialized in `save`. I'm not sure scikit-learn models are pickle-friendly or not. The safest way is remove `self.model = None` from `__init__`. This is considered as best-practice in Jina development.

Got it, thanks for the pointer.

> Btw, could you please add some unit test, e.g. check if the transformed `ndarray` fits into the expected dimensions and are non-zero, non-NaN.

I'm aware that these unit tests are crucial for any development, however this is my first open-source contribution and PR and I am very new to these tools. Would it be possible for you to send me any guides/resources where I can learn how to build these tests? I understand what needs to be done, just don't know how to do it. Thanks!

**Comment by hanxiao** [comment]
see `tests/`

**Comment by anish2197** [comment]
Hi Han, I've written some unit tests for the new encoders. They seem to be working. Got some "Arrays are not equal" assertion errors, but I realised some encoders need to be initialised with random_state for reproducible results. Please review the changes. Thanks.

**Comment by anish2197** [comment]
recheckcla

**Comment by anish2197** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by anish2197** [comment]
> * Create a new class `FitTransformEncoder`, with `__init__(self, output_dim, random_state)`, and `encode()`

Where should this class be defined?

**Comment by hanxiao** [comment]
You can put all classes in one `numeric/fit_transform.py`. `FitTransformEncoder` and all its inheritance shall be in this file. This file structure is considered as the best practice.

File: tests/executors/encoders/numeric/__init__.py [hunk 1]
```diff
@@ -56,6 +56,8 @@ def test_save_and_load(self):
         self.assertTrue(os.path.exists(encoder.save_abspath))
         encoder_loaded = BaseExecutor.load(encoder.save_abspath)
         encoded_data_test = encoder_loaded.encode(test_data)
+        print(encoded_data_test.shape)
```
Comment (JoanFM) [inline tests/executors/encoders/numeric/__init__.py:59]: Please remove prints

**Comment by JoanFM** [comment]
Hey @anish2197 ,

You would need to realign on master, and do not forget to move ur tests under tests/unit since they have been moved since this PR was opened.

**Comment by JoanFM** [comment]
Hey @anish2197.

These are some good features. Could you please realign to master so that the tests can pass? We can maybe provide some assistance?

**Comment by anish-pawar** [comment]
Hi Joan, sorry for this late reply. I was extremely busy and couldn't work on this Jina PR. Could you please send me the steps to realign to master?

**Comment by JoanFM** [comment]
> Hi Joan, sorry for this late reply. I was extremely busy and couldn't work on this Jina PR. Could you please send me the steps to realign to master?

No worries, I just consider it to be a useful PR and was wondering if we could move forward. 

As per the instructions, you can follow from: https://stackoverflow.com/questions/16329776/how-to-keep-a-git-branch-in-sync-with-master#:~:text=Whenever%20you%20want%20to%20get,HEAD%3Amaster%20.

1 git fetch from your feature branch (make sure the feature branch you are working on is update to date)

2 git rebase origin master

3 if any conflict shall arise, resolve them one by one

4 use git rebase --continue once all conflicts are dealt with

5 git push --force

Let me know if you need any help with it.

**Comment by hanxiao** [comment]
i will address the problem in a separate PR

**Comment by JoanFM** [comment]
> i will address the problem in a separate PR

u mean the one related to model fitting the batch everytime?
